### PR TITLE
Make transmuting inhabited to uninhabited types an error

### PR DESCRIPTION
**Pull request by canndrew**
This changes the behaviour of the transmute intrinsic so that, for any
uninhabited type `Void` and any inhabited type `NonVoid`:
- `transmute::<Void, T>` is okay
- `transmute::<NonVoid, Void>` is an error

### CODE REVIEW

**Review by nagisa — APPROVED** [review]
> transmute::<Void, T> is okay

This can’t be possibly called, can it? Good thing you cannot get function pointer of an intrinsic, as making one up for this particular function would be kinda tricky :)

Overall this patch LGTM.

**Comment by oli-obk** [comment]
the issue code should probably be used as a regression test.

**Comment by durka** [comment]
@nagisa you can still construct a void by using `transmute_copy` or pointer casts (both seem to correctly invoke UB), so it could still be called.

**Comment by arielb1** [comment]
I think we should rather be changing the code in `trans::block` to make transmuting to a `!` result in an `unreachable`, rather than hackily adding errors.

**Comment by canndrew** [comment]
> At the `ty::layout` level, I would imagine we want to propagate this further.

The hackyness begins at whatever level we start pretending that uninhabited types are ZSTs, so I agree that the lower we can push that the better. What exactly do you have in mind though?

> Why are you making `transmute::<Void, NonVoid>` legal?

Why not? It's always safe to transmute from an uninhabited type to anything else (every element of `Void` has the same size as `T` for any `T`). Plus, `transmute::<A, B>` isn't a symmetrical operation w.r.t. the types `A` and `B` so it's not necessarily more elegant to make it symmetrical w.r.t. their sizes. I wouldn't feel strongly about changing it though.

**Comment by alexcrichton** [comment]
ping, what's the status of this?

**Comment by canndrew** [comment]
@alexcrichton It prolly needs the lang team to have a look at it and see if they agree with it. Other than that though it should be good to merge.

**Comment by steveklabnik** [comment]
/cc @rust-lang/lang 

Also, there's no reviewer for this? I'm not sure who should be assigned.

**Comment by eddyb** [comment]
The patch is alright, I suppose, but I don't understand the motivation between this change.
AFAICT the main effect is making `transmute::<ZST, Void>` an error (which requires at least a crater run).

**Comment by canndrew** [comment]
> AFAICT the main effect is making `transmute::<ZST, Void>` an error

Correct. Because it's an insane thing to do and it has the same problems as `uninitialized::<Void>`. That said, I'm starting to feel we should settle the questions around `uninitialized::<Void>` first before merging anything like this :/

**Comment by eddyb** [comment]
I definitely want a bunch of warnings in this area, not so sure about outright errors.
Deny-by-default lints are quite good at this sort of bad-unsafe-code sort of thing IMO.

**Comment by alexcrichton** [comment]
If we're thinking of holding off on this until we've settled more questions about uninitialized types, perhaps we should close in the meantime? (to help clear out the queue)

**Comment by canndrew** [comment]
@alexcrichton Righto. Closed for now.

**Pull request closed without merging**
